src/shamir.rs

```diff
@@ -0,0 +1,364 @@
+//!
+//! This module contains the implementation of [`Shamir secret sharing`] using finite fields.
+//! It uses FF implementations provided by [`ff`] crate and is mostly useful when working with
+//! prime fields as they support arithmetic addition and multiplication of shares that is used
+//! to implement secure multiplication protocol.
+//!
+//! [`Shamir secret sharing`](https://en.wikipedia.org/wiki/Shamir%27s_Secret_Sharing)
+//! [`ff`](https://crates.io/crates/ff)
+//!
+use ff::{Field, PrimeField};
+use rand_core::RngCore;
+use std::iter::repeat_with;
+use std::num::NonZeroU8;
+use thiserror::Error;
+
+/// Shamir secret sharing
+pub struct SecretSharing {
+    /// Threshold
+    k: u8,
+
+    /// Number of shares produced by the `share` method
+    n: u8,
+}
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("Wrong or insecure secret sharing scheme. (expected {k} >= {n} > 1)")]
+    BadSharingScheme { k: u8, n: u8 },
+    #[error("Not enough shares to reconstruct secret: need at least {actual}, got {required}")]
+    NotEnoughShares { required: u8, actual: u8 },
+    #[error("Prime field element inversion failed: {v:?}")]
+    InvertError { v: Box<[u8]> },
+}
+
+/// Represents a single share: (x, f(x)) point
+#[derive(Clone)]
+pub struct Share<F> {
+    x: u8,
+    y: F,
+}
+
+#[derive(Debug, Error)]
+pub enum ShareError {
+    #[error("Can't perform operation on shares as they don't have the same X coordinate: ({lhs} != {rhs})")]
+    MisalignedShares { lhs: u8, rhs: u8 },
+}
+
+impl SecretSharing {
+    /// Constructs a new instance, returning `ShamirError` if input values do not allow
+    /// to construct a valid Shamir secret sharing scheme.
+    ///
+    /// # Errors
+    /// Rejects the following inputs:
+    /// * k == 1
+    /// * k > n
+    pub fn new(k: NonZeroU8, n: NonZeroU8) -> Result<Self, Error> {
+        if k.get() == 1 || k > n {
+            Err(Error::BadSharingScheme {
+                k: k.get(),
+                n: n.get(),
+            })
+        } else {
+            Ok(Self {
+                k: k.get(),
+                n: n.get(),
+            })
+        }
+    }
+
+    /// Minimum number of shares required to reconstruct a secret
+    #[must_use]
+    pub fn threshold(&self) -> u8 {
+        self.k
+    }
+
+    /// Consumes and splits a given secret into `n` shares. only `k` <= `n` shares is required
+    /// to reconstruct it later
+    pub fn split<R: RngCore, F: PrimeField>(&self, secret: F, rng: R) -> Vec<Share<F>> {
+        // generate polynomial of k-1 degree
+        let coefficients = gen_polynomial::<R, F>(rng)
+            .take(self.k as usize)
+            .collect::<Vec<_>>();
+
+        let mut shares = Vec::with_capacity(self.n as usize);
+
+        // sample n points by evaluating the polynomial
+        for x in 1..=self.n {
+            let x_i = F::from(u64::from(x));
+            let mut y = secret;
+            for c in &coefficients {
+                y += (*c) * x_i;
```

Comment:
I am almost certain you want:
  y += (*c) * x_i.pow(j)

Where j is the index of the loop

Reply:
Maybe you can build in reverse and avoid the "pow" call.  The usual pattern here is to start from the non-constant coefficients and do something like

```rust
let mut y = 0;
for c in coefficients.iter().rev().chain([secret].iter()) {
  y *= x_i;
  y += *c;
}
```

Reply:
Or more simply...

```rust
let mut y = 0;
for c in coefficients.iter().rev() {
  y += *c;
  y *= x_i;
}
y += secret;
```

Reply:
yea, thanks @benjaminsavage for spotting a bug here - certainly a miss. The funny part that there is another bug at the line 81 where I take `k` coefficients while I actually need `k-1` and that issue finally appeared after I fixed polynomial degree.

@martinthomson I like what you suggested, in the future we may switch to use [`pow_vartime`](https://docs.rs/ff/0.12.0/ff/trait.Field.html#method.pow_vartime) as potentially more efficient